dashboard/app/controllers/admin_users_controller.rb

```diff
@@ -36,4 +36,23 @@ def confirm_email
     end
   end
 
+  def undelete_user_form
+  end
+
+  def undelete_user
+    user = User.with_deleted.where(id: params[:user_id].to_i).first
+
+    if user
+      if user.deleted?
+        user.restore
+        user.save!
+        flash[:alert] = 'User undeleted!'
+      else
+        flash[:alert] = 'User was never deleted!'
+      end
+    else
+      flash[:alert] = 'User not found!'
+    end
+    render :undelete_user_form
```

Comment:
it's a common pattern to redirect to the undelete_user_form action (instead of rendering) on success (for one, this prevents the form from being submitted again on reload)